File: scripts/http-fetch/fetch_retry.py
"""fetch_retry.py — zero-dependency HTTP fetch with per-attempt timeout,
retry-with-backoff, and a typed error. Python stdlib only (urllib) — no requests.

Runtime requirement: Python 3.8+. No pip dependencies.

    from fetch_retry import fetch_retry, FetchRetryError
    resp = fetch_retry("https://example.com", retries=3)
    body = resp.read().decode()

CLI demo:  python fetch_retry.py https://example.com
"""

from __future__ import annotations

import random
import sys
import time
import urllib.error
import urllib.request
from email.utils import parsedate_to_datetime
from datetime import datetime, timezone
from http.client import HTTPResponse
from typing import Callable, Mapping, Optional

# Status codes worth retrying — transient server / rate-limit conditions.
RETRYABLE_STATUS = frozenset({429, 500, 502, 503, 504})


class FetchRetryError(Exception):
    """Raised when every attempt fails. Carries the HTTP status (if any),
    the number of attempts made, and the underlying cause."""

    def __init__(
        self,
        message: str,
        *,
        status: Optional[int] = None,
        attempts: int,
        url: str,
        cause: Optional[BaseException] = None,
    ) -> None:
        super().__init__(message)
        self.status = status
        self.attempts = attempts
        self.url = url
        self.__cause__ = cause


def _parse_retry_after(value: Optional[str]) -> Optional[float]:
    """Parse a Retry-After header (delta-seconds or HTTP-date) into seconds."""
    if not value:
        return None
    value = value.strip()
    if value.isdigit():
        return max(0.0, float(value))
    try:
        when = parsedate_to_datetime(value)
    except (TypeError, ValueError):
        return None
    if when is None:
        return None
    if when.tzinfo is None:
        when = when.replace(tzinfo=timezone.utc)
    return max(0.0, (when - datetime.now(timezone.utc)).total_seconds())


def _backoff_with_jitter(attempt: int, base: float, cap: float) -> float:
    """Exponential backoff (base * 2**(attempt-1)) with full jitter, capped."""
    exp = min(cap, base * (2 ** (attempt - 1)))
    return random.uniform(0.0, exp)  # full jitter: [0, exp)


def fetch_retry(
    url: str,
    *,
    method: str = "GET",
    data: Optional[bytes] = None,
    headers: Optional[Mapping[str, str]] = None,
    retries: int = 3,
    timeout: float = 10.0,
    backoff: float = 0.3,
    max_backoff: float = 20.0,
    on_retry: Optional[Callable[[int, BaseException, float], None]] = None,
) -> HTTPResponse:
    """Open ``url`` with a per-attempt timeout and exponential backoff + jitter.

    Retries on network errors and retryable status codes (429, 500, 502, 503,
    504), respecting ``Retry-After`` when present. Does NOT retry other 4xx —
    those are raised immediately (wrapped) so the caller can inspect them. On
    final failure raises :class:`FetchRetryError`.

    Returns the live ``HTTPResponse`` (call ``.read()`` to get the body).

    :param retries: max retries *after* the first attempt (default 3 → up to 4).
    :param timeout: per-attempt timeout in seconds.
    """
    req = urllib.request.Request(url, data=data, method=method)
    for key, val in (headers or {}).items():
        req.add_header(key, val)

    last_error: Optional[BaseException] = None
    last_status: Optional[int] = None

    for attempt in range(1, retries + 2):
        delay: Optional[float] = None
        try:
            return urllib.request.urlopen(req, timeout=timeout)  # noqa: S310
        except urllib.error.HTTPError as err:
            # The server answered with an error status.
            last_status = err.code
            last_error = err
            if err.code not in RETRYABLE_STATUS:
                # Non-retryable (e.g. 400/401/403/404) — surface immediately.
                raise FetchRetryError(
                    f"HTTP {err.code} {err.reason}",
                    status=err.code,
                    attempts=attempt,
                    url=url,
                    cause=err,
                ) from err
            delay = _parse_retry_after(err.headers.get("Retry-After"))
        except (urllib.error.URLError, OSError) as err:
            # Network failure or timeout (socket.timeout is an OSError subclass).
            last_error = err

        if attempt > retries:
            break
        if delay is None:
            delay = _backoff_with_jitter(attempt, backoff, max_backoff)
        if on_retry is not None and last_error is not None:
            on_retry(attempt, last_error, delay)
        time.sleep(delay)

    raise FetchRetryError(
        f"Failed after {retries + 1} attempt(s): {last_error}",
        status=last_status,
        attempts=retries + 1,
        url=url,
        cause=last_error,
    )


# --- CLI demo -------------------------------------------------------------
def _main(argv: list) -> int:
    url = argv[1] if len(argv) > 1 else "https://example.com"

    def _log(attempt: int, error: BaseException, delay: float) -> None:
        print(f"  retry {attempt} after {delay:.2f}s — {error}", file=sys.stderr)

    try:
        resp = fetch_retry(url, retries=3, on_retry=_log)
        body = resp.read()
        print(f"✓ {resp.status} {resp.reason} — {len(body)} bytes")
        return 0
    except FetchRetryError as err:
        print(
            f"✗ {err} (status={err.status}, attempts={err.attempts})",
            file=sys.stderr,
        )
        return 1


if __name__ == "__main__":
    sys.exit(_main(sys.argv))

File: scripts/http-fetch/fetch-retry.ts
// fetch-retry.ts — zero-dependency HTTP fetch with per-attempt timeout,
// retry-with-backoff, and a typed error. Built-ins only.
//
// Runtime requirement: Node 18+ (global `fetch`, `AbortSignal.timeout`).
// No npm dependencies.
//
//   import { fetchRetry, FetchRetryError } from './fetch-retry';
//   const res = await fetchRetry('https://example.com', { retries: 3 });
//
// CLI demo:  npx tsx fetch-retry.ts https://example.com

/** Status codes worth retrying — transient server / rate-limit conditions. */
const RETRYABLE_STATUS = new Set([429, 500, 502, 503, 504]);

export interface FetchRetryOptions extends RequestInit {
  /** Max retries *after* the first attempt. Default 3 (so up to 4 requests). */
  retries?: number;
  /** Per-attempt timeout in ms. Default 10_000. */
  timeoutMs?: number;
  /** Base backoff in ms; grows exponentially per attempt. Default 300. */
  backoffMs?: number;
  /** Cap on a single backoff sleep in ms. Default 20_000. */
  maxBackoffMs?: number;
  /** Hook for observing each retry (logging/metrics). */
  onRetry?: (info: { attempt: number; error: Error; delayMs: number }) => void;
}

/** Thrown when every attempt fails. Carries the HTTP status (if any) and count. */
export class FetchRetryError extends Error {
  readonly status?: number;
  readonly attempts: number;
  readonly url: string;
  override readonly cause?: unknown;

  constructor(
    message: string,
    opts: { status?: number; attempts: number; url: string; cause?: unknown },
  ) {
    super(message);
    this.name = 'FetchRetryError';
    this.status = opts.status;
    this.attempts = opts.attempts;
    this.url = opts.url;
    this.cause = opts.cause;
  }
}

const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

/**
 * Parse a `Retry-After` header (delta-seconds or HTTP-date) into ms.
 * Returns undefined if absent or unparseable.
 */
function parseRetryAfter(value: string | null): number | undefined {
  if (!value) return undefined;
  const secs = Number(value);
  if (Number.isFinite(secs)) return Math.max(0, secs * 1000);
  const when = Date.parse(value);
  if (!Number.isNaN(when)) return Math.max(0, when - Date.now());
  return undefined;
}

/**
 * fetch() with a per-attempt timeout and exponential backoff + jitter.
 *
 * Retries on network errors and retryable status codes (429, 500, 502, 503,
 * 504), respecting `Retry-After` when present. Does NOT retry other 4xx — those
 * are returned as-is for the caller to handle. On final failure throws
 * {@link FetchRetryError}.
 */
export async function fetchRetry(
  url: string,
  options: FetchRetryOptions = {},
): Promise<Response> {
  const {
    retries = 3,
    timeoutMs = 10_000,
    backoffMs = 300,
    maxBackoffMs = 20_000,
    onRetry,
    signal: callerSignal,
    ...init
  } = options;

  let lastError: Error | undefined;
  let lastStatus: number | undefined;

  for (let attempt = 1; attempt <= retries + 1; attempt++) {
    // Per-attempt timeout, combined with any caller-provided signal.
    const timeoutSignal = AbortSignal.timeout(timeoutMs);
    const signal = callerSignal
      ? AbortSignal.any([callerSignal, timeoutSignal])
      : timeoutSignal;

    let response: Response;
    try {
      response = await fetch(url, { ...init, signal });
    } catch (err) {
      // Network failure or timeout (AbortError). Retry unless caller aborted.
      lastError = err instanceof Error ? err : new Error(String(err));
      if (callerSignal?.aborted) {
        throw new FetchRetryError('Request aborted by caller', {
          attempts: attempt,
          url,
          cause: lastError,
        });
      }
      if (attempt > retries) break;
      const delayMs = backoffWithJitter(attempt, backoffMs, maxBackoffMs);
      onRetry?.({ attempt, error: lastError, delayMs });
      await sleep(delayMs);
      continue;
    }

    if (response.ok || !RETRYABLE_STATUS.has(response.status)) {
      // Success, or a non-retryable status (e.g. 400/401/403/404) — hand back.
      return response;
    }

    // Retryable status code.
    lastStatus = response.status;
    lastError = new Error(`HTTP ${response.status} ${response.statusText}`);
    if (attempt > retries) break;

    const retryAfter = parseRetryAfter(response.headers.get('retry-after'));
    const delayMs =
      retryAfter ?? backoffWithJitter(attempt, backoffMs, maxBackoffMs);
    onRetry?.({ attempt, error: lastError, delayMs });
    // Drain the body so the connection can be reused.
    await response.body?.cancel().catch(() => {});
    await sleep(delayMs);
  }

  throw new FetchRetryError(
    `Failed after ${retries + 1} attempt(s): ${lastError?.message ?? 'unknown error'}`,
    { status: lastStatus, attempts: retries + 1, url, cause: lastError },
  );
}

/** Exponential backoff (base * 2^(attempt-1)) with full jitter, capped. */
function backoffWithJitter(attempt: number, base: number, cap: number): number {
  const exp = Math.min(cap, base * 2 ** (attempt - 1));
  return Math.random() * exp; // full jitter: [0, exp)
}

// --- CLI demo -------------------------------------------------------------
// `npx tsx fetch-retry.ts <url>` — fetches a URL and prints status + size.
if (import.meta.url === `file://${process.argv[1]}`) {
  const url = process.argv[2] ?? 'https://example.com';
  fetchRetry(url, {
    retries: 3,
    onRetry: ({ attempt, error, delayMs }) =>
      console.error(
        `  retry ${attempt} after ${Math.round(delayMs)}ms — ${error.message}`,
      ),
  })
    .then(async (res) => {
      const body = await res.text();
      console.log(`✓ ${res.status} ${res.statusText} — ${body.length} bytes`);
    })
    .catch((err: unknown) => {
      if (err instanceof FetchRetryError) {
        console.error(
          `✗ ${err.message} (status=${err.status ?? 'n/a'}, attempts=${err.attempts})`,
        );
      } else {
        console.error('✗ unexpected error:', err);
      }
      process.exit(1);
    });
}
